build/result.go

```diff
@@ -22,166 +20,246 @@ import (
 	"golang.org/x/sync/errgroup"
 )
 
-func NewResultContext(ctx context.Context, c *client.Client, solveOpt client.SolveOpt, res *gateway.Result) (*ResultContext, error) {
-	def, err := getDefinition(ctx, res)
-	if err != nil {
-		return nil, err
-	}
-	return getResultAt(ctx, c, solveOpt, def, nil)
-}
-
-func getDefinition(ctx context.Context, res *gateway.Result) (*result.Result[*pb.Definition], error) {
-	return result.ConvertResult(res, func(ref gateway.Reference) (*pb.Definition, error) {
-		st, err := ref.ToState()
-		if err != nil {
-			return nil, err
-		}
-		def, err := st.Marshal(ctx)
-		if err != nil {
-			return nil, err
-		}
-		return def.ToPB(), nil
-	})
-}
-
-func getResultAt(ctx context.Context, c *client.Client, solveOpt client.SolveOpt, targets *result.Result[*pb.Definition], statusChan chan *client.SolveStatus) (*ResultContext, error) {
-	ctx, cancel := context.WithCancel(ctx)
-	defer cancel()
+// NewResultContext wraps a call to client.Build, additionally returning a
+// ResultContext alongside the standard response and error.
+//
+// This ResultContext can be used to execute build steps in the same context as
+// the build occurred, which can allow easy debugging of build failures and
+// successes.
+//
+// If the returned ResultContext is not nil, the caller must call Done() on it.
+func NewResultContext(ctx context.Context, cc *client.Client, opt client.SolveOpt, product string, buildFunc gateway.BuildFunc, ch chan *client.SolveStatus) (*ResultContext, *client.SolveResponse, error) {
```

Comment:
It's quite confusing that a function that actually runs the build is called `NewResultContext` (this can be follow-up)

Reply:
>  I really want to avoid another function that we call Build,

If that's what it does, then it should be called like it. We need to draw a chart of these new components/functions to see if the design makes sense and what is the the correct scope for each pkg/function.

This is not `Wrap`. Wrap is something that takes a build callback and returns a new build callback. This one takes a build callback and calls it.

Reply:
Have renamed `ResultContext` to `ResultHandle` as suggested in slack, but will leave the name of `NewResultHandle` for a follow-up. I think we can do some other refactoring of the `Build` family of functions in Buildx that will make this a little easier.